proto/substrait/algebra.proto

```diff
@@ -228,6 +228,90 @@ message AggregateRel {
+  // But WindowRelFunction remove the WindowSpec property, because all the WindowRelFunctions
+  // share the same WindowSpec() in WindowRel.
+  message WindowRelFunction {
+    // Points to a function_anchor defined in this plan, which must refer
+    // to a window function in the associated YAML file. Required; 0 is
+    // considered to be a valid anchor/reference.
+    uint32 function_reference = 1;
+
+    // The arguments to be bound to the function. This must have exactly the
+    // number of arguments specified in the function definition, and the
+    // argument types must also match exactly:
+    //
+    //  - Value arguments must be bound using FunctionArgument.value, and
+    //    the expression in that must yield a value of a type that a function
+    //    overload is defined for.
+    //  - Type arguments must be bound using FunctionArgument.type, and a
+    //    function overload must be defined for that type.
+    //  - Enum arguments must be bound using FunctionArgument.enum
+    //    followed by Enum.specified, with a string that case-insensitively
+    //    matches one of the allowed options.
+    repeated FunctionArgument arguments = 9;
+
+    // Options to specify behavior for corner cases, or leave behavior
+    // unspecified if the consumer does not need specific behavior in these
+    // cases.
+    repeated FunctionOption options = 11;
+
+    // Must be set to the return type of the function, exactly as derived
+    // using the declaration in the extension.
+    Type output_type = 7;
+
+    // Describes which part of the window function to perform within the
+    // context of distributed algorithms. Required. Must be set to
+    // INITIAL_TO_RESULT for window functions that are not decomposable.
+    AggregationPhase phase = 6;
+
+    // Specifies whether equivalent records are merged before being aggregated.
+    // Optional, defaults to AGGREGATION_INVOCATION_ALL.
+    AggregateFunction.AggregationInvocation invocation = 10;
+
+    // Defines the record relative to the current record from which the window
+    // extends. The bound is inclusive. If the lower bound indexes a record
+    // greater than the upper bound, TODO (null range/no records passed?
+    // wrapping around as if lower/upper were swapped? error? null?).
+    // Optional; defaults to the start of the partition.
+    Expression.WindowFunction.Bound lower_bound = 5;
+
+    // Defines the record relative to the current record up to which the window
+    // extends. The bound is inclusive. If the upper bound indexes a record
+    // less than the lower bound, TODO (null range/no records passed?
+    // wrapping around as if lower/upper were swapped? error? null?).
+    // Optional; defaults to the end of the partition.
+    Expression.WindowFunction.Bound upper_bound = 4;
+
+    // Defines the bounds type: ROWS, RANGE
+    BoundsType bounds_type = 12;
+  }
+
+  enum BoundsType {
```

Comment:
minor: can we make this enum a top-level enum like `AggregationPhase`.

This will make it easier/nicer to re-use in the `WindowFunction` expression, where it is also needed.

Reply:
This has been moved into `WindowFunction` as having all the references in `WindowRelFunction` use the same prefix seemed the most consistent. 
```proto
    ...
    Expression.WindowFunction.Bound lower_bound = 5;
    ...
    Expression.WindowFunction.Bound upper_bound = 4;
    ...
    Expression.WindowFunction.BoundsType bounds_type = 12;
```